Failing tests are:
`tests/database.py::test_nodes_to_dataframe_simple - AssertionError: DataFrame.columns are different`
and
`tests/iotable.py::test_iotable_nodes_to_dataframe - AssertionError: DataFrame.columns are different`

The previous `SqliteExtDatabase` class seems to return fields in the order defined in the class, while `SqliteDatabase` returns in the database-internal order. I'm not sure if the order here is actually important. If not, `check_like=True` could be added to `assert_frame_equal` here - waiting for feedback on this one.

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -17,7 +17,6 @@ on:
       - ".gitignore"
   pull_request:
     branches: [main, develop]
-    types: [opened, reopened] # excludes syncronize to avoid redundant trigger from commits on PRs
     paths-ignore:
       - "**.md"
       - "**.bib"

--- a/bw2data/backends/base.py
+++ b/bw2data/backends/base.py
@@ -1209,10 +1209,14 @@ Here are the type values usually used for nodes:
 
         """
         if columns is None:
-            # Feels like magic
-            df = pandas.DataFrame(self)
+            rows = [dict(obj) for obj in self]
+            # Build DataFrame with deterministic column ordering, independent of ORM field order.
+            ordered_columns = sorted({key for row in rows for key in row})
+            df = pandas.DataFrame(rows, columns=ordered_columns)
         else:
-            df = pandas.DataFrame([{field: obj.get(field) for field in columns} for obj in self])
+            rows = [{field: obj.get(field) for field in columns} for obj in self]
+            # Keep user-defined column order as-is.
+            df = pandas.DataFrame(rows, columns=columns)
         if return_sorted:
             sort_columns = ["name", "reference product", "location", "unit"]
             df = df.sort_values(by=[column for column in sort_columns if column in df.columns])

--- a/bw2data/search/indices.py
+++ b/bw2data/search/indices.py
@@ -13,16 +13,18 @@ class IndexManager:
     def __init__(self, database_path):
         self.path = os.path.join(projects.request_directory("search"), database_path)
         self.db = SqliteDatabase(self.path)
-        if not os.path.exists(self.path) or len(self.db.get_tables()) == 0:
-            self.create()
+        with self.db.connection_context():
+            if not os.path.exists(self.path) or len(self.db.get_tables()) == 0:
+                self.create()
 
     def get(self):
         return self
 
     def create(self):
         self.delete_database()
-        with self.db.bind_ctx(MODELS):
-            self.db.create_tables(MODELS)
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                self.db.create_tables(MODELS)
 
     def _format_dataset(self, ds):
         def _fix_location(string):
@@ -52,34 +54,38 @@ class IndexManager:
 
     def add_datasets(self, datasets):
         all_dataset = list(datasets)
-        with self.db.bind_ctx(MODELS):
-            for chunk_range in range(0, len(datasets), 100):
-                for model in MODELS:
-                    model.insert_many(
-                        [
-                            self._format_dataset(ds)
-                            for ds in all_dataset[chunk_range : chunk_range + 100]
-                        ]
-                    ).execute()
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                for chunk_range in range(0, len(datasets), 100):
+                    for model in MODELS:
+                        model.insert_many(
+                            [
+                                self._format_dataset(ds)
+                                for ds in all_dataset[chunk_range : chunk_range + 100]
+                            ]
+                        ).execute()
 
     def update_dataset(self, ds):
-        with self.db.bind_ctx(MODELS):
-            for model in MODELS:
-                model.delete().where(
-                    model.code == ds["code"], model.database == ds["database"]
-                ).execute()
-                model.insert(**self._format_dataset(ds)).execute()
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                for model in MODELS:
+                    model.delete().where(
+                        model.code == ds["code"], model.database == ds["database"]
+                    ).execute()
+                    model.insert(**self._format_dataset(ds)).execute()
 
     def delete_dataset(self, ds):
-        with self.db.bind_ctx(MODELS):
-            for model in MODELS:
-                model.delete().where(
-                    model.code == ds["code"], model.database == ds["database"]
-                ).execute()
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                for model in MODELS:
+                    model.delete().where(
+                        model.code == ds["code"], model.database == ds["database"]
+                    ).execute()
 
     def delete_database(self):
-        with self.db.bind_ctx(MODELS):
-            self.db.drop_tables(MODELS)
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                self.db.drop_tables(MODELS)
 
     def close(self):
         self.db.close()
@@ -113,26 +119,27 @@ class IndexManager:
                 DeprecationWarning,
             )
 
-        with self.db.bind_ctx(MODELS):
-            if string == "*":
-                query = BW2Schema
-            else:
-                query = BW2Schema.search_bm25(
-                    self.escape_search_for_fts5(string),
-                    weights=weights,
+        with self.db.connection_context():
+            with self.db.bind_ctx(MODELS):
+                if string == "*":
+                    query = BW2Schema
+                else:
+                    query = BW2Schema.search_bm25(
+                        self.escape_search_for_fts5(string),
+                        weights=weights,
+                    )
+                return list(
+                    query.select(
+                        BW2Schema.name,
+                        BW2Schema.comment,
+                        BW2Schema.product,
+                        BW2Schema.categories,
+                        BW2Schema.synonyms,
+                        BW2Schema.location,
+                        BW2Schema.database,
+                        BW2Schema.code,
+                    )
+                    .limit(limit)
+                    .dicts()
+                    .execute()
                 )
-            return list(
-                query.select(
-                    BW2Schema.name,
-                    BW2Schema.comment,
-                    BW2Schema.product,
-                    BW2Schema.categories,
-                    BW2Schema.synonyms,
-                    BW2Schema.location,
-                    BW2Schema.database,
-                    BW2Schema.code,
-                )
-                .limit(limit)
-                .dicts()
-                .execute()
-            )

--- a/bw2data/serialization.py
+++ b/bw2data/serialization.py
@@ -38,13 +38,16 @@ class JsonWrapper:
     @classmethod
     def load(self, file):
         if anyjson:
-            return anyjson.deserialize(open(file, encoding="utf-8").read())
+            with open(file, encoding="utf-8") as f:
+                return anyjson.deserialize(f.read())
         else:
-            return json.load(open(file, encoding="utf-8"))
+            with open(file, encoding="utf-8") as f:
+                return json.load(f)
 
     @classmethod
     def load_bz2(self, filepath):
-        return JsonWrapper.loads((bz2.BZ2File(filepath).read()).decode("utf-8"))
+        with bz2.BZ2File(filepath) as f:
+            return JsonWrapper.loads(f.read().decode("utf-8"))
 
     @classmethod
     def dumps(self, data):
@@ -240,7 +243,8 @@ class PickledDict(SerializedDict):
 
     def deserialize(self):
         try:
-            return self.unpack(pickle.load(open(self.filepath, "rb")))
+            with open(self.filepath, "rb") as f:
+                return self.unpack(pickle.load(f))
         except ImportError:
             TEXT = "Pickle deserialization error in file '%s'" % self.filepath
             raise PickleError(TEXT)

--- a/bw2data/tests.py
+++ b/bw2data/tests.py
@@ -12,6 +12,21 @@ from bw2data import config, databases, geomapping, methods
 from bw2data.parameters import parameters
 from bw2data.project import projects
 
+def _close_sqlite_handles():
+    """Close all known substitutable SQLite handles to avoid descriptor leaks."""
+    try:
+        if not projects.db.db.is_closed():
+            projects.db.db.close()
+    except Exception:
+        pass
+
+    for _, substitutable_db in config.sqlite3_databases:
+        try:
+            if not substitutable_db.db.is_closed():
+                substitutable_db.db.close()
+        except Exception:
+            pass
+
 
 class BW2DataTest(unittest.TestCase):
     _multiprocess_can_split_ = True
@@ -19,7 +34,9 @@ class BW2DataTest(unittest.TestCase):
     def setUp(self):
         config.dont_warn = True
         config.is_test = True
+        _close_sqlite_handles()
         tempdir = Path(tempfile.mkdtemp())
+        self._tempdir = tempdir
         project_name = "".join(random.choices(string.ascii_lowercase, k=18))
         projects.change_base_directories(
             base_dir=tempdir,
@@ -32,6 +49,10 @@ class BW2DataTest(unittest.TestCase):
 
     def extra_setup(self):
         pass
+
+    def tearDown(self):
+        _close_sqlite_handles()
+        shutil.rmtree(self._tempdir, ignore_errors=True)
 
     def test_setup_clean(self):
         self.assertEqual(list(databases), [])
@@ -47,6 +68,7 @@ class BW2DataTest(unittest.TestCase):
 def bw2test(wrapped, instance, args, kwargs):
     config.dont_warn = True
     config.is_test = True
+    _close_sqlite_handles()
     tempdir = Path(tempfile.mkdtemp())
     project_name = "".join(random.choices(string.ascii_lowercase, k=18))
     projects.change_base_directories(

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,5 +1,6 @@
 """Fixtures for bw2data"""
 
+import gc
 import sqlite3
 
 import pytest
@@ -23,7 +24,26 @@ def num_revisions():
 
 @pytest.fixture(autouse=True, scope="function")
 def close_database_handle_on_exit_to_avoid_windows_test_errors():
-    from bw2data.backends import sqlite3_lci_db
+    from bw2data import config
+    from bw2data.project import projects
 
+    def close_handles():
+        for _, substitutable_db in config.sqlite3_databases:
+            try:
+                if not substitutable_db.db.is_closed():
+                    substitutable_db.db.close()
+            except Exception:
+                pass
+
+        try:
+            if not projects.db.db.is_closed():
+                projects.db.db.close()
+        except Exception:
+            pass
+
+        # Encourage cleanup of objects holding stale SQLite handles.
+        gc.collect()
+
+    close_handles()
     yield
-    sqlite3_lci_db.db.close()
+    close_handles()

